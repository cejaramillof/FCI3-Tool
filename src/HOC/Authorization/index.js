import React from "react";
import { connect } from 'react-redux'

const emptyComponent = () => <span></span>;

const Authorization = (
  roles = [],
  GreenComponent = emptyComponent,
  RedComponent = emptyComponent
) => {

  const AuthWrapper = props => {
    const isAuthorized = roles.indexOf(props.user.role) >= 0;
    return (
      <>
        {isAuthorized
          ? <GreenComponent {...props} />
          : <RedComponent {...props} />
        }
      </>
    )
  }

  const mapStateToProps = state => ({ user: state.user });
  return connect(mapStateToProps)(AuthWrapper)
}

export default Authorization