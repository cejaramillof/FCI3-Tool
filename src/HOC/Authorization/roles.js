export const ADMIN = 'ADMIN';
export const RECRUITER = 'RECRUITER';
export const GATEKEEPER = 'GATEKEEPER';
export const ALL= [ADMIN, RECRUITER, GATEKEEPER];