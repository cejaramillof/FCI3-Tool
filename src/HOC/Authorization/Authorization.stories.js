import React from 'react';
import { withKnobs } from "@storybook/addon-knobs";
import centered from '@storybook/addon-centered/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Authorization from './index';
import * as ROLES from './roles';
import readme from './readme.md';
import 'index.scss';

export default {
  title: 'HOC/Authorization ',
  parameters: {
    notes: readme,
  },
  component: Authorization,
  decorators: [withKnobs, centered]
};

// Redux mock store setup
const middlewares = []
const mockStore = configureStore(middlewares)
const initialState = {
  user: {
    role: ROLES.ADMIN
  }
}
const store = mockStore(initialState)

const ProviderWrapper = props => (
  <Provider store={store}>
    {props.children}
  </Provider>
)
//

const greenComponent = () => <div>Admins are Authorice to see this</div>;
const AuthComponent = Authorization([ROLES.ADMIN], greenComponent)


export const Admin = () => (
  <ProviderWrapper>
    <h1>Authorization HOC</h1>
    <p>Use Authorization Higher Order Component to enable a comonent depending of the user role</p>
    <div><AuthComponent /></div>
  </ProviderWrapper>
);

