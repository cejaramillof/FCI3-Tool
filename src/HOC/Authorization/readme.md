**HOC**

# Authorization


Use Authorization **H**igher **O**rder **C**omponent to enable a comonent depending of the user role

```jsx 
import Authorization from 'Authorization';
const AuthComponent = Authorization(<List-of-User-Roles>, <component>,<component>);
<AuthComponent/> 
```
**Authorization** HOC is connected to redux store
``` const mapStateToProps = state => ({ user: state.user }); ```

Uses ```user.role``` to get current user role to validate
Params:

| Param             | Type        |                                 |          |
| ----------------- |:-----------:| :-------------------------------| :--------|
| roles             | Array       | List of Authorized roles        | Optional |
| GreenComponent    | Component   | Component to render if Auth     | Optional |
| RedComponent      | Component   | Component to render if NOT Auth | Optional |

Example using **Roles Constants** for only Recruiters

```jsx 
import Authorization from 'Authorization';
import * as ROLES from 'Authorization/roles';

const requiredAuthComponent = props => <p>{props.children}</p>;

const AuthComponent = Authorization([ROLES.RECRUITER], requiredAuthComponent);

<AuthComponent>This component is authorised for this user role</AuthComponent> 
```

Example using **Roles Constants** for Recruiters and Gatekeepers

```jsx 
import Authorization from 'Authorization';
import * as ROLES from 'Authorization/roles';

const requiredAuthComponent = props => <p>{props.children}</p>;

const AuthComponent = Authorization([ROLES.RECRUITER, ROLES.GATEKEEPER], requiredAuthComponent);

<AuthComponent>This component is authorised for this user role</AuthComponent> 
```


Example using **Roles Constants** for all user Roles

```jsx 
import Authorization from 'Authorization';
import * as ROLES from 'Authorization/roles';

const requiredAuthComponent = props => <p>{props.children}</p>;

const AuthComponent = Authorization(ROLES.ALL, requiredAuthComponent);

<AuthComponent>This component is authorised for this user role</AuthComponent> 
```

