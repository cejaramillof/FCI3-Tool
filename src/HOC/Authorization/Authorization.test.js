import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import Authorization from './index';
import * as ROLES from './roles';

// Redux mock store setup
const middlewares = []
const mockStore = configureStore(middlewares)
const initialState = {
  user: {
    role: ROLES.ADMIN
  }
}
const store = mockStore(initialState)

const notRender = () => <div id="not-redered">Should not be rendered</div>;
const emptyComponent = props => <div data-role={props.role}>{props.children}</div>;
const AdminComponent = Authorization([ROLES.ADMIN], emptyComponent)
const GatekeeperComponent = Authorization([ROLES.GATEKEEPER], notRender, emptyComponent)

export const Wrapper = mount(
  <Provider store={store}>
    <AdminComponent role="admin-only">lorem ipsum</AdminComponent>
    <GatekeeperComponent role="gatekeeper-only">Only Gatekeepers</GatekeeperComponent>
  </Provider>
)

describe('Authorization HOC', function () {
  it('Should render correctly', () => expect(Wrapper).toMatchSnapshot());
  it('Should render Admin only', () => expect(Wrapper.find({ 'data-role': 'admin-only' })).toHaveLength(1));
  it('Should render Gatekeeper second component', () => expect(Wrapper.find({ 'data-role': 'gatekeeper-only' })).toHaveLength(1));
  it('Should not render #not-redered element', () => expect(Wrapper.find('#not-redered')).toHaveLength(0));
});