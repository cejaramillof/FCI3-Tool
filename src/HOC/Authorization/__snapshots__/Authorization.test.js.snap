// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization HOC Should render correctly 1`] = `ReactWrapper {}`;
